runners/core-java/src/main/java/org/apache/beam/runners/core/SimpleDoFnRunner.java

```diff
@@ -765,13 +747,37 @@ private void verifyAbsoluteTimeDomain() {
       }
     }
 
+    /**
+     * <ul>Ensures that:
+     *   <li>Users can't set {@code outputTimestamp} for processing time timers.</li>
+     *   <li>Event time timers' {@code outputTimestamp} is set before window expiration.</li></ul>
+     */
+    private void setAndVerifyOutputTimestamp() {
+      // Output timestamp is currently not supported in processing time timers.
+      if (outputTimestamp != null && !TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
+        throw new IllegalStateException("Cannot set outputTimestamp in processing time domain.");
+      }
+      // Output timestamp is set to the delivery time if not initialized by an user.
+      if (outputTimestamp == null) {
+        outputTimestamp = target;
+      }
+
+      System.out.println(outputTimestamp);
+      if (TimeDomain.EVENT_TIME.equals(spec.getTimeDomain())) {
+        Instant windowExpiry = window.maxTimestamp().plus(allowedLateness);
+        checkArgument(!outputTimestamp.isAfter(windowExpiry),
```

Comment:
This may be an even stricter check (probably should be even) - we should enforce that the outputTimestamp is no later than the `maxTimestamp` of the window, if I understand windows properly

Reply:
Done.